fs/zbd_zenfs.cc

```diff
@@ -525,192 +502,302 @@ IOStatus ZonedBlockDevice::AllocateMetaZone(Zone **out_meta_zone) {
+        if (!full) PutActiveIOZoneToken();
+      } else {
+        IOStatus release_status = z->CheckRelease();
+        if (!release_status.ok()) return release_status;
       }
-      IOStatus status = z->CheckRelease();
-      if (!status.ok()) return status;
     }
   }
-  return Status::OK();
+  return IOStatus::OK();
 }
 
-IOStatus ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime,
-                                        Zone **out_zone) {
-  Zone *allocated_zone = nullptr;
-  Zone *finish_victim = nullptr;
-  unsigned int best_diff = LIFETIME_DIFF_NOT_GOOD;
-  int new_zone = 0;
-  IOStatus s;
-  bool ok = false;
-  (void)ok;
-  ZenFSMetricsLatencyGuard guard(
-      metrics_,
-      IOType::kUnknown == IOType::kWAL
-          ? ZENFS_LABEL_DETAILED(IO_ALLOC, WAL, LATENCY)
-          : ZENFS_LABEL_DETAILED(IO_ALLOC, NON_WAL, LATENCY),
-      Env::Default());
-  metrics_->ReportQPS(ZENFS_LABEL(IO_ALLOC, QPS), 1);
-
-  *out_zone = nullptr;
+void ZonedBlockDevice::WaitForOpenIOZoneToken(bool prioritized) {
+  long allocator_open_limit;
 
-  // Check if a deferred IO error was set
-  s = GetZoneDeferredStatus();
-  if (!s.ok()) {
-    return s;
+  /* Avoid non-priortized allocators from starving prioritized ones */
+  if (prioritized) {
+    allocator_open_limit = max_nr_open_io_zones_;
+  } else {
+    allocator_open_limit = max_nr_open_io_zones_ - 1;
   }
 
-  io_zones_mtx.lock();
-
-  /* Make sure we are below the zone open limit */
-  {
-    std::unique_lock<std::mutex> lk(zone_resources_mtx_);
-    zone_resources_.wait(lk, [this] {
-      if (open_io_zones_.load() < max_nr_open_io_zones_) return true;
+  /* Wait for an open IO Zone token - after this function returns
+   * the caller is allowed to write to a closed zone. The callee
+   * is responsible for calling a PutOpenIOZoneToken to return the resource
+   */
+  std::unique_lock<std::mutex> lk(zone_resources_mtx_);
+  zone_resources_.wait(lk, [this, allocator_open_limit] {
+    if (open_io_zones_.load() < allocator_open_limit) {
+      open_io_zones_++;
+      return true;
+    } else {
       return false;
-    });
+    }
+  });
+}
+
+bool ZonedBlockDevice::GetActiveIOZoneTokenIfAvailable() {
+  /* Grap an active IO Zone token if available - after this function returns
+   * the caller is allowed to write to a closed zone. The callee
+   * is responsible for calling a PutActiveIOZoneToken to return the resource
+   */
+  std::unique_lock<std::mutex> lk(zone_resources_mtx_);
+  if (active_io_zones_.load() < max_nr_active_io_zones_) {
+    active_io_zones_++;
+    return true;
   }
+  return false;
+}
 
-  /* Reset any unused zones and finish used zones under capacity treshold*/
-  for (const auto z : io_zones) {
-    if (!z->Acquire()) {
-      continue;
-    }
+void ZonedBlockDevice::PutOpenIOZoneToken() {
+  std::unique_lock<std::mutex> lk(zone_resources_mtx_);
+  open_io_zones_--;
+  zone_resources_.notify_one();
```

Comment:
Why?

Reply:
@MaisenbacherD I don't think that matters: https://en.cppreference.com/w/cpp/thread/condition_variable
```text
The thread that intends to modify the shared variable has to

    acquire a std::mutex (typically via std::lock_guard)
    perform the modification while the lock is held
    execute notify_one or notify_all on the std::condition_variable (the lock does not need to be held for notification) 
```

Reply:
In fact:
```text
The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock. However, some implementations (in particular many implementations of pthreads) recognize this situation and avoid this "hurry up and wait" scenario by transferring the waiting thread from the condition variable's queue directly to the queue of the mutex within the notify call, without waking it up. 
```

So, in some implementations, it is slower to hold the lock when calling `notify_one`.

Reply:
Because `zone_resource_mtx_` is otherwise still locked while waking up the waiting thread that wants to lock `zone_resources_mtx_` as well, but can't and goes directly to sleep again.
This is also explained in the notes section of: 
https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

Reply:
It does not hurt, so i'll fold this change in